e2e-tests/helpers/boundary-api-client.ts

```diff
@@ -0,0 +1,358 @@
+
+      /** {@link https://developer.hashicorp.com/boundary/docs/concepts/domain-model/session-recordings Session Recordings Documentation} */
+      SessionRecording: new BoundaryApiClient.SessionRecordingServiceApi(
+        openapiConfiguration,
+      ),
+
+      /** {@link https://developer.hashicorp.com/boundary/docs/concepts/domain-model/sessions Sessions Documentation} */
+      Session: new BoundaryApiClient.SessionServiceApi(openapiConfiguration),
+
+      /** {@link https://developer.hashicorp.com/boundary/docs/concepts/domain-model/storage-buckets Storage Buckets Documentation} */
+      StorageBucket: new BoundaryApiClient.StorageBucketServiceApi(
+        openapiConfiguration,
+      ),
+
+      /** {@link https://developer.hashicorp.com/boundary/docs/concepts/domain-model/targets Targets Documentation} */
+      Target: new BoundaryApiClient.TargetServiceApi(openapiConfiguration),
+
+      /** {@link https://developer.hashicorp.com/boundary/docs/concepts/domain-model/users Users Documentation} */
+      User: new BoundaryApiClient.UserServiceApi(openapiConfiguration),
+
+      /** {@link https://developer.hashicorp.com/boundary/api-docs/worker-service Worker Service Api Docs} */
+      Worker: new BoundaryApiClient.WorkerServiceApi(openapiConfiguration),
+    };
+
+    return clients;
+  }
+
+  // it's assumed that ids are unique so the type isn't needed to be known when
+  // iterating over created resources
+  skipCleanup(identifiable: { id: string }): void {
+    if (!identifiable.id) {
+      throw new Error('The `id` field is expected on the resource');
+    }
+
+    this.skipCleanupResources.push(identifiable);
+  }
+
+  private get openapiConfiguration() {
+    const { controllerAddr, createdResources: resources } = this;
+
+    const captureCreatedResourcesMiddleware = {
+      async post(context) {
+        const resourceWasCreated =
+          context.response.status === BoundaryApiStatusCodes.created &&
+          context.response.headers.get('content-type') ===
+            BOUNDARY_API_CREATED_RESOURCE_HTTP_CONTENT_TYPE;
+
+        // if the respronse doesn't represent a created resource, return early
+        if (!resourceWasCreated) {
+          return context.response;
+        }
+
+        // the url is the being used to determine the resource type, it returns the type in pluralized form
+        const pluralizedResourceType = extractPluralizedResourceTypeFromUrl(
+          context.response.url,
+        );
+        // the response `json` can only be read once, and needs to be read later by the generated api client,
+        // instead we can clone the response and this allows us to read the `json` on the cloned response only
+        const json = await context.response.clone().json();
+
+        // it's assume that json response has an `id` in order for it to be a created resource
+        if (pluralizedResourceType && typeof json?.id === 'string') {
+          resources[pluralizedResourceType] ??= [];
```

Comment:
TIL about the [nullish coalescing assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_assignment), neat